### PR TITLE
Migrate table scan related code to `DAGStorageInterpreter`

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #4118

Problem Summary:

### What is changed and how it works?
- remove `DAGQueryBlock` used in `DAGStorageInterpreter`
- move `executeRemoteQuery`, `executeCastAfterTableScan` and `executePushedDownFilter` from `DAGQueryBlockInterpreter` to `DAGStorageInterpreter`

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Coprocessor/DAGStorageInterpreter.h [hunk 1]
```diff
@@ -50,9 +48,9 @@ class DAGStorageInterpreter
 public:
     DAGStorageInterpreter(
         Context & context_,
-        const DAGQueryBlock & query_block_,
         const TiDBTableScan & table_scan,
-        const std::vector<const tipb::Expr *> & conditions_,
+        const String & pushed_down_filter_executor_id_,
```
Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/DAGStorageInterpreter.h:52]: Maybe can wrap infomation about filter push down into a struct.

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 2]
```diff
@@ -169,70 +150,10 @@ AnalysisResult analyzeExpressions(
...
-        if (!table_regions_info.local_regions.empty() || !table_regions_info.remote_regions.empty())
-        {
-            has_region_to_read = true;
-            break;
-        }
-    }
-    if (!has_region_to_read)
-        throw TiFlashException(fmt::format("Dag Request does not have region to read for table: {}", table_scan.getLogicalTableID()), Errors::Coprocessor::BadRequest);
     // construct pushed down filter conditions.
     std::vector<const tipb::Expr *> conditions;
```
Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:158]: Maybe wrap preparations of filter pushdown into a function.

File: dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp [hunk 3]
```diff
@@ -165,16 +314,159 @@ void DAGStorageInterpreter::execute(DAGPipeline & pipeline)
...
+        for (size_t i = 0; i < after_cast_source_columns.size(); ++i)
+        {
+            project_for_remote_read.emplace_back(original_source_columns[i].name, after_cast_source_columns[i].name);
+        }
+        assert(!project_for_remote_read.empty());
+        assert(pipeline.streams_with_non_joined_data.empty());
+        assert(remote_read_streams_start_index <= pipeline.streams.size());
+        size_t i = 0;
+        // local streams
+        while (i < remote_read_streams_start_index)
```
Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:383]: nit: Maybe extract the function into two part. The first one handle cast of local streams, the second one handle cast of remote streams.

File: dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp [hunk 4]
```diff
@@ -165,16 +314,159 @@ void DAGStorageInterpreter::execute(DAGPipeline & pipeline)
...
+        }
+    }
+}
+
+void DAGStorageInterpreter::executeRemoteQuery(DAGPipeline & pipeline)
+{
+    assert(!remote_requests.empty());
+    DAGSchema & schema = remote_requests[0].schema;
+#ifndef NDEBUG
+    auto schema_match = [&schema](const DAGSchema & other) {
```
Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:409]: Why not move the schemaMatch function defined in DAGQueryBlockInterpreter into a util file? And reuse it.

File: dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp (hunk 3 above)
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:374]: I think it's better to put it in a function...

File: dbms/src/Flash/Coprocessor/RemoteRequest.cpp [hunk 5]
```diff
@@ -36,13 +42,14 @@ RemoteRequest RemoteRequest::build(const RegionRetryList & retry_regions, DAGCon
     DAGSchema schema;
     tipb::DAGRequest dag_req;
     auto * executor = dag_req.mutable_root_executor();
-    if (selection != nullptr)
+
+    if (push_down_filter.hasValue())
```
Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/RemoteRequest.cpp:46]: nit: How about extract the logic into PushDownFilter，the function name can be PushDownFilter::buildSelection

File: dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp [hunk 6]
```diff
@@ -127,28 +136,159 @@ MakeRegionQueryInfos(
...
+    pipeline.transform([&](auto & stream) { table_scan_io_input_streams.push_back(stream); });
+
+    if (pipeline.streams.empty())
+    {
+        pipeline.streams.emplace_back(null_stream_if_empty);
+        // reset remote_read_streams_start_index for null_stream_if_empty.
+        remote_read_streams_start_index = 1;
+    }
+
+    /// We don't wants to the table to be dropped during the lifetime of this query,
```
Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:265]:
```suggestion
    /// We don't want the table to be dropped during the lifetime of this query,
```

Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:265]:
```suggestion
    /// We don't want the table to be dropped during the lifetime of this query,
```

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 7]
```diff
@@ -169,223 +151,16 @@ AnalysisResult analyzeExpressions(
...
-    if (!has_region_to_read)
-        throw TiFlashException(fmt::format("Dag Request does not have region to read for table: {}", table_scan.getLogicalTableID()), Errors::Coprocessor::BadRequest);
-    // construct pushed down filter conditions.
-    std::vector<const tipb::Expr *> conditions;
-    if (query_block.selection)
-    {
-        for (const auto & condition : query_block.selection->selection().conditions())
-            conditions.push_back(&condition);
-    }
+    auto push_down_filter = PushDownFilter::toPushDownFilter(query_block.selection);
```
Comment (JaySon-Huang) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:158]:
```suggestion
    const auto push_down_filter = PushDownFilter::toPushDownFilter(query_block.selection);
```

File: dbms/src/Flash/Coprocessor/RemoteRequest.cpp [hunk 8]
```diff
@@ -36,15 +42,8 @@ RemoteRequest RemoteRequest::build(const RegionRetryList & retry_regions, DAGCon
...
-    {
-        executor->set_tp(tipb::ExecType::TypeSelection);
-        executor->set_executor_id(selection->executor_id());
-        auto * new_selection = executor->mutable_selection();
-        for (const auto & condition : selection->selection().conditions())
-            *new_selection->add_conditions() = condition;
-        executor = new_selection->mutable_child();
-    }
+
+    executor = push_down_filter.constructSelectionForRemoteRead(executor);
```
Comment (JaySon-Huang) [inline dbms/src/Flash/Coprocessor/RemoteRequest.cpp:46]:
```suggestion
    auto * executor = push_down_filter.constructSelectionForRemoteRead(dag_req.mutable_root_executor());
```
nit: maybe merge L44 and L46?

File: dbms/src/Flash/Coprocessor/RemoteRequest.cpp (hunk 5 above)
Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/RemoteRequest.cpp]: ok

**Review by ywqzzy — APPROVED** [review]
LGTM

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**